# Tutorial 4 Python Notebook 

In this tutorial there are tasks for working with lists, arrays, dictionaries, and data frames.

* Lists = Task 1-3
* Arrays = Task 4-7 
* Dictionaries = Task 8-10
* Data frames = Task 11-16

The aim of this tutorial notebook is to give you some (guided) hands-on experience working with different data structures in Python.

In [3]:
import numpy as np 
import pandas as pd

## Lists 

### Task 1

Create a list containing strings, numbers, a list, and Boolean values. 


In [ ]:
## your answer 


#### Task 1 solution 

There are endless potential solutions, the primary point is to emphasize that lists can contain multiple data types and to check your understanding of what these data types are by name. 


In [4]:
list_data = ["orange", "apple", [5, 7, 9, 11], True, 125.17, 75.83]

list_data

['orange', 'apple', [5, 7, 9, 11], True, 125.17, 75.83]

### Task 2

How would you index to the first item in the nested list in the list you created in task 1?

In [ ]:
## your answer 



#### Task 2 solution

This will vary depending on your answer to Task 1. But using `list_data`, to access the number 5 we need to use the following code

In [5]:
list_data[2][0]

5

### Task 3

Convert the list below into a 1-dimensional array. 

In [7]:
list1 = [12.23, 13.32, 100.1, 36.45]

print(list1)

[12.23, 13.32, 100.1, 36.45]


<details><summary style='color:lightblue'>HINT: Where do we get the array data structure? CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

Remember, arrays are **not** a built-in Python data structure. We access them through the `NumPy` package. 
  

#### Task 3 solution 

In [8]:
array_l1 = np.array(list1) 

print(array_l1) # notice the array print out does not include commas between values as in the list data structure above 

print(type(array_l1))

[ 12.23  13.32 100.1   36.45]
<class 'numpy.ndarray'>


## Array 

## Task 4

Use the keyboard variable below to spell your name by referencing the letters in the array. 

<details><summary style='color:lightblue'>HINT: Example solution. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

brittany is: 
    
`print(keyboard[2,4], keyboard[0,3], keyboard[0,7], keyboard[0,4], keyboard[0,4], keyboard[1,0], keyboard[2,5], keyboard[0,5])`

In [10]:
keyboard = np.array([ ['q','w','e','r','t','y','u','i','o','p'],
                      ['a','s','d','f','g','h','j','k','l',';'],
                      ['z','x','c','v','b','n','m','<','>','?']
                   ])

In [16]:
## your answer here 
print(keyboard[2][4].capitalize() + keyboard[0][2] + keyboard[2][5])


Ben


#### Task 4 solution 

This will of course vary dependin on your name (and I do not think anyone on the course shares my name)

In [15]:
print(keyboard[2,4], keyboard[0,3], keyboard[0,7], keyboard[0,4], keyboard[0,4], keyboard[1,0], keyboard[2,5], keyboard[0,5])

b r i t t a n y


In [17]:
## if you want the letters formatted closer together we can use the '+' operator to add then together without a space 
print(keyboard[2,4] + keyboard[0,3] + keyboard[0,7] + keyboard[0,4] + keyboard[0,4] + keyboard[1,0] + keyboard[2,5] + keyboard[0,5])

brittany


### Task 5 

Create a 1-dimensional array with values ranging from 20 and 80.

In [20]:
## your answer 
one_dim_array = np.arange(20, 81)
one_dim_array

array([20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,
       37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70,
       71, 72, 73, 74, 75, 76, 77, 78, 79, 80])

#### Task 5 Solution 

In [19]:
array2 = np.arange(20,81)

array2

array([20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36,
       37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,
       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70,
       71, 72, 73, 74, 75, 76, 77, 78, 79, 80])

### Task 6

Reverse the array created in Task 5.

<details><summary style='color:lightblue'>HINT: A reminder. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>
    
Indexing with arrays (and lists and dataframes) can be done with the following syntax ` my_array[start_index : stop_index : step/jump]`

<details><summary style='color:lightblue'>HINT: Another reminder. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

Remember that reverse indexing in Python and R are very different! 


In [21]:
## your answer 
array2[::-1]

array([80, 79, 78, 77, 76, 75, 74, 73, 72, 71, 70, 69, 68, 67, 66, 65, 64,
       63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47,
       46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30,
       29, 28, 27, 26, 25, 24, 23, 22, 21, 20])

#### Task 6 solution 

In [22]:
array2_rev = array2[::-1] # reminder, the first colon : here means `everything` 

print(array2_rev)

[80 79 78 77 76 75 74 73 72 71 70 69 68 67 66 65 64 63 62 61 60 59 58 57
 56 55 54 53 52 51 50 49 48 47 46 45 44 43 42 41 40 39 38 37 36 35 34 33
 32 31 30 29 28 27 26 25 24 23 22 21 20]


### Task 7

Create a 2D array with 1 across the border and 0 inside/middle.

<details><summary style='color:lightblue'>HINT: Breaking down the problem. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

This may sound easy enough, but is actually quite tricky as we are working in 2 dimensions! As a first step, create a 2-D array filled with ones (there is a specific function for this). Then replace the inside or middle values of the array with 0.

In [26]:
## your answer 
twod_array = np.ones((5,5))
twod_array[1:-1, 1:-1] = 0
twod_array

array([[1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1.]])

#### Task 7 solution 

In [27]:
# step 1 create an array with ones 
array3 = np.ones((5,5), dtype = 'int')

print(array3)

[[1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]
 [1 1 1 1 1]]


In [28]:
## next change the values in the middle to be 0 using indexing
array3[1:-1, 1:-1] = 0

print(array3)

[[1 1 1 1 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 0 0 0 1]
 [1 1 1 1 1]]


## Dictionaries

### Task 8

Create a dictionary comprised of colors blue, red, green; animals dog, cat, horse; and age 33, 56, 24.

In [ ]:
## your answer 


#### Task 8  Solution 

In [29]:
dict1 = {"colors": ["blue", "red", "green"],
         "animals": ["dog", "cat", "horse"],
         "age": [33, 56, 24]}

print(dict1)

{'colors': ['blue', 'red', 'green'], 'animals': ['dog', 'cat', 'horse'], 'age': [33, 56, 24]}


### Task 9

Add a new key value pair to the list created in Task 8 called flower comprised of daisy, rose, and lily. 

In [30]:
## your answer 
dict1['flower'] = ['rose', 'lily', 'tulip']

#### Task 9 Solution 

In [ ]:
dict1["flower"] = ["daisy", "rose", "lily"]

print(dict1)

### Task 10

Convert your dictionary into a data frame. 

<details><summary style='color:lightblue'>HINT: Where do we get the data frame data structure? CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

Remember, data frames are **not** a built-in Python data structure. We access them through the `pandas` package. 
  

In [31]:
## your answer 
df = pd.DataFrame.from_dict(dict1)

#### Task 10 Solution 

In [32]:
df_dict = pd.DataFrame.from_dict(dict1)

df_dict

colors animals  age flower
0   blue     dog   33   rose
1    red     cat   56   lily
2  green   horse   24  tulip

## Data frames 

For this series of tasks we will be using a publicly available dataset from [Public Health Scotland around Stroke Activity](https://www.opendata.nhs.scot/dataset/scottish-stroke-statistics/resource/47656572-e196-40c8-83e8-08b0b223b2e6). This dataset provides "Information on hospital activity related to cerebrovascular disease (including stroke and subarachnoid haemorrhage)." Look through the link and read the data dictionary at the bottom to familiarise yourself with the variables.  

We will first read in the data. You can read in data from a URL with the `read_csv()` function but inputting the URL as a character string - how helpful! From the link above to the data set there is a URL I have copied at the top of the page. It is good practice when first reading in a data set to name it and add `_raw` or some delineation that it is the raw data. As you process the data for your analytic purposes, you can then save the data in an object without this delineation. This allows you to maintain an object with a version of the raw data that you can refer to later if needed. 

In [33]:
stroke_raw = pd.read_csv("https://www.opendata.nhs.scot/dataset/f5dcf382-e6ca-49f6-b807-4f9cc29555bc/resource/47656572-e196-40c8-83e8-08b0b223b2e6/download/stroke_activitybyhbr.csv")

In [34]:
# it is always a good idea to do a quick visual check of data once you read it in to spot an obvious or blantant parsing issues 
stroke_raw # looks good! 

FinancialYear        HBR HBRQF AdmissionType AdmissionTypeQF  \
0           2013/14  S08000015   NaN           All               d   
1           2013/14  S08000015   NaN           All               d   
2           2013/14  S08000015   NaN           All               d   
3           2013/14  S08000015   NaN           All               d   
4           2013/14  S08000015   NaN           All               d   
...             ...        ...   ...           ...             ...   
43195       2022/23  S92000003     d      Transfer             NaN   
43196       2022/23  S92000003     d      Transfer             NaN   
43197       2022/23  S92000003     d      Transfer             NaN   
43198       2022/23  S92000003     d      Transfer             NaN   
43199       2022/23  S92000003     d      Transfer             NaN   

            AgeGroup AgeGroupQF      Sex SexQF                   Diagnosis  \
0         0-44 years        NaN    Males   NaN     Cerebrovascular Disease   
1         0-44 years        NaN  Females   NaN     Cerebrovascular Disease   
2        45-64 years        NaN    Males   NaN     Cerebrovascular Disease   
3        45-64 years        NaN  Females   NaN     Cerebrovascular Disease   
4        65-74 years        NaN    Males   NaN     Cerebrovascular Disease   
...              ...        ...      ...   ...                         ...   
43195    45-64 years        NaN      All     d  TIAs and related syndromes   
43196    65-74 years        NaN      All     d  TIAs and related syndromes   
43197   75plus years        NaN      All     d  TIAs and related syndromes   
43198  under75 years        NaN      All     d  TIAs and related syndromes   
43199            All          d      All     d  TIAs and related syndromes   

       NumberOfDischarges NumberOfDischargesQF    CrudeRate CrudeRateQF  \
0                    57.0                  NaN    61.547764         NaN   
1                    42.0                  NaN    44.180298         NaN   
2                   231.0                  NaN   445.773832         NaN   
3                   209.0                  NaN   371.238765         NaN   
4                   244.0                  NaN  1198.899371         NaN   
...                   ...                  ...          ...         ...   
43195               262.0                  NaN    17.582418         NaN   
43196               300.0                  NaN    50.371236         NaN   
43197               726.0                  NaN   151.792976         NaN   
43198               572.0                  NaN    11.436302         NaN   
43199              1298.0                  NaN    23.686564         NaN   

              EASR  
0        66.515711  
1        43.481241  
2       444.876027  
3       372.420079  
4      1229.931638  
...            ...  
43195    16.915867  
43196    50.648669  
43197   153.689663  
43198    10.970167  
43199    23.814921  

[43200 rows x 15 columns]

In [35]:
# using pd.read_csv() will import the data for us as a pandas data frame 
type(stroke_raw)

pandas.core.frame.DataFrame

### Task 11 

Look through the imported data to check it looks as it should based on the data dictionary. Are all the expected variables included? What dimensions does the dataframe have? Do the dtypes of these variables look to be correct? 

<details><summary style='color:lightblue'>HINT: A useful new function! CLICK HERE TO SEE</summary>
    
There is a useful function called `head()` which will print by default the first 5 rows of a dataframe. The counterpart is `tail()` which print the last 5 rows by default. Both functions take the argument `n = ` if you wish to specify a different number of rows other than 5. There are counterpart functions in R with the same name and functionality!
    
This will help in solving the task, but you will need to use some other summary/description functions as well.

In [ ]:
## your answer 


#### Task 11 Solution

In [36]:
stroke_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43200 entries, 0 to 43199
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   FinancialYear         43200 non-null  object 
 1   HBR                   43200 non-null  object 
 2   HBRQF                 2880 non-null   object 
 3   AdmissionType         43200 non-null  object 
 4   AdmissionTypeQF       10800 non-null  object 
 5   AgeGroup              43200 non-null  object 
 6   AgeGroupQF            7200 non-null   object 
 7   Sex                   43200 non-null  object 
 8   SexQF                 14400 non-null  object 
 9   Diagnosis             43200 non-null  object 
 10  NumberOfDischarges    39503 non-null  float64
 11  NumberOfDischargesQF  3697 non-null   object 
 12  CrudeRate             39503 non-null  float64
 13  CrudeRateQF           3697 non-null   object 
 14  EASR                  43200 non-null  float64
dtypes: float64(3), obje

In [37]:
stroke_raw.describe(include = 'all')

FinancialYear        HBR HBRQF AdmissionType AdmissionTypeQF  \
count          43200      43200  2880         43200           10800   
unique            10         15     1             4               1   
top          2013/14  S08000015     d           All               d   
freq            4320       2880  2880         10800           10800   
mean             NaN        NaN   NaN           NaN             NaN   
std              NaN        NaN   NaN           NaN             NaN   
min              NaN        NaN   NaN           NaN             NaN   
25%              NaN        NaN   NaN           NaN             NaN   
50%              NaN        NaN   NaN           NaN             NaN   
75%              NaN        NaN   NaN           NaN             NaN   
max              NaN        NaN   NaN           NaN             NaN   

          AgeGroup AgeGroupQF    Sex  SexQF                Diagnosis  \
count        43200       7200  43200  14400                    43200   
unique           6          1      3      1                        4   
top     0-44 years          d  Males      d  Cerebrovascular Disease   
freq          7200       7200  14400  14400                    10800   
mean           NaN        NaN    NaN    NaN                      NaN   
std            NaN        NaN    NaN    NaN                      NaN   
min            NaN        NaN    NaN    NaN                      NaN   
25%            NaN        NaN    NaN    NaN                      NaN   
50%            NaN        NaN    NaN    NaN                      NaN   
75%            NaN        NaN    NaN    NaN                      NaN   
max            NaN        NaN    NaN    NaN                      NaN   

        NumberOfDischarges NumberOfDischargesQF     CrudeRate CrudeRateQF  \
count         39503.000000                 3697  39503.000000        3697   
unique                 NaN                    1           NaN           1   
top                    NaN                    c           NaN           c   
freq                   NaN                 3697           NaN        3697   
mean            314.787434                  NaN    220.864123         NaN   
std            1284.948500                  NaN    538.363106         NaN   
min               0.000000                  NaN      0.000000         NaN   
25%               2.000000                  NaN      1.664115         NaN   
50%              25.000000                  NaN     24.363294         NaN   
75%             167.000000                  NaN    175.604957         NaN   
max           35410.000000                  NaN   5856.146694         NaN   

                EASR  
count   43200.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean      231.331289  
std       545.092634  
min         0.000000  
25%         2.602546  
50%        28.850160  
75%       187.272256  
max      5895.829863

In [38]:
stroke_raw.shape

(43200, 15)

In [39]:
stroke_raw.ndim # 2 dimensions as expected - we are working with a dataframe after all!

2

In [40]:
stroke_raw.head()

FinancialYear        HBR HBRQF AdmissionType AdmissionTypeQF     AgeGroup  \
0       2013/14  S08000015   NaN           All               d   0-44 years   
1       2013/14  S08000015   NaN           All               d   0-44 years   
2       2013/14  S08000015   NaN           All               d  45-64 years   
3       2013/14  S08000015   NaN           All               d  45-64 years   
4       2013/14  S08000015   NaN           All               d  65-74 years   

  AgeGroupQF      Sex SexQF                Diagnosis  NumberOfDischarges  \
0        NaN    Males   NaN  Cerebrovascular Disease                57.0   
1        NaN  Females   NaN  Cerebrovascular Disease                42.0   
2        NaN    Males   NaN  Cerebrovascular Disease               231.0   
3        NaN  Females   NaN  Cerebrovascular Disease               209.0   
4        NaN    Males   NaN  Cerebrovascular Disease               244.0   

  NumberOfDischargesQF    CrudeRate CrudeRateQF         EASR  
0                  NaN    61.547764         NaN    66.515711  
1                  NaN    44.180298         NaN    43.481241  
2                  NaN   445.773832         NaN   444.876027  
3                  NaN   371.238765         NaN   372.420079  
4                  NaN  1198.899371         NaN  1229.931638

In [41]:
stroke_raw.tail(n = 6)

FinancialYear        HBR HBRQF AdmissionType AdmissionTypeQF  \
43194       2022/23  S92000003     d      Transfer             NaN   
43195       2022/23  S92000003     d      Transfer             NaN   
43196       2022/23  S92000003     d      Transfer             NaN   
43197       2022/23  S92000003     d      Transfer             NaN   
43198       2022/23  S92000003     d      Transfer             NaN   
43199       2022/23  S92000003     d      Transfer             NaN   

            AgeGroup AgeGroupQF  Sex SexQF                   Diagnosis  \
43194     0-44 years        NaN  All     d  TIAs and related syndromes   
43195    45-64 years        NaN  All     d  TIAs and related syndromes   
43196    65-74 years        NaN  All     d  TIAs and related syndromes   
43197   75plus years        NaN  All     d  TIAs and related syndromes   
43198  under75 years        NaN  All     d  TIAs and related syndromes   
43199            All          d  All     d  TIAs and related syndromes   

       NumberOfDischarges NumberOfDischargesQF   CrudeRate CrudeRateQF  \
43194                10.0                  NaN    0.342946         NaN   
43195               262.0                  NaN   17.582418         NaN   
43196               300.0                  NaN   50.371236         NaN   
43197               726.0                  NaN  151.792976         NaN   
43198               572.0                  NaN   11.436302         NaN   
43199              1298.0                  NaN   23.686564         NaN   

             EASR  
43194    0.337105  
43195   16.915867  
43196   50.648669  
43197  153.689663  
43198   10.970167  
43199   23.814921

There are 15 variables in our dataset, as expected, and 43200 observations. Of these variables, there are 12 which are characters and may be better suited as categorical.

### Task 12 

We do not need all of the columns in the dataset. The only variables we need for the next tasks are `FinancialYear`, `AdmissionType`, `AgeGroup`, `Diagnosis`, and `NumberOfDischarges`. Within the `HBR` variable, "S92000003" is the country code for Scotland. The `Sex` variable we do not need for this task, but it too includes an aggregate level "All". Filter the data such that only these aggregate level are included for these 2 variables. Filter the data accordingly and save this processed dataset into an object called `stroke`.

<details><summary style='color:lightblue'>HINT 1: Think about your data structures! CLICK HERE TO SEE</summary>
Remember to think about your input data structures - for example, perhaps you will want to use a list to list the columns we want to keep

<details><summary style='color:lightblue'>HINT 2: Indexing operators CLICK HERE TO SEE</summary>

We learned about `loc` and `iloc` this week to indexing location based on name or numeric index, respectively

<details><summary style='color:lightblue'>HINT 3: Resulting data frame dimensions. CLICK HERE TO SEE</summary>

the object `stroke` should contain 960 rows and 5 columns 


#### Task 12 solution 

In [42]:
stroke = stroke_raw.loc[(stroke_raw["HBR"] == "S92000003") & 
                       (stroke_raw["Sex"] == "All")].loc[:,["FinancialYear", "AdmissionType", 
                                                            "AgeGroup", "Diagnosis",
                                                            "NumberOfDischarges"]]

stroke

FinancialYear AdmissionType       AgeGroup                   Diagnosis  \
7284        2013/14           All     0-44 years     Cerebrovascular Disease   
7285        2013/14           All    45-64 years     Cerebrovascular Disease   
7286        2013/14           All    65-74 years     Cerebrovascular Disease   
7287        2013/14           All   75plus years     Cerebrovascular Disease   
7288        2013/14           All  under75 years     Cerebrovascular Disease   
...             ...           ...            ...                         ...   
43195       2022/23      Transfer    45-64 years  TIAs and related syndromes   
43196       2022/23      Transfer    65-74 years  TIAs and related syndromes   
43197       2022/23      Transfer   75plus years  TIAs and related syndromes   
43198       2022/23      Transfer  under75 years  TIAs and related syndromes   
43199       2022/23      Transfer            All  TIAs and related syndromes   

       NumberOfDischarges  
7284               1239.0  
7285               6822.0  
7286               6608.0  
7287              14331.0  
7288              14669.0  
...                   ...  
43195               262.0  
43196               300.0  
43197               726.0  
43198               572.0  
43199              1298.0  

[960 rows x 5 columns]

In [43]:
## you could also use isin instead of == to filter HBR and Sex
stroke_raw.loc[(stroke_raw["HBR"].isin(["S92000003"])) & 
              stroke_raw["Sex"].isin(["All"])].loc[:,["FinancialYear", "AdmissionType",
                                                      "AgeGroup", "Diagnosis",
                                                      "NumberOfDischarges"]]

FinancialYear AdmissionType       AgeGroup                   Diagnosis  \
7284        2013/14           All     0-44 years     Cerebrovascular Disease   
7285        2013/14           All    45-64 years     Cerebrovascular Disease   
7286        2013/14           All    65-74 years     Cerebrovascular Disease   
7287        2013/14           All   75plus years     Cerebrovascular Disease   
7288        2013/14           All  under75 years     Cerebrovascular Disease   
...             ...           ...            ...                         ...   
43195       2022/23      Transfer    45-64 years  TIAs and related syndromes   
43196       2022/23      Transfer    65-74 years  TIAs and related syndromes   
43197       2022/23      Transfer   75plus years  TIAs and related syndromes   
43198       2022/23      Transfer  under75 years  TIAs and related syndromes   
43199       2022/23      Transfer            All  TIAs and related syndromes   

       NumberOfDischarges  
7284               1239.0  
7285               6822.0  
7286               6608.0  
7287              14331.0  
7288              14669.0  
...                   ...  
43195               262.0  
43196               300.0  
43197               726.0  
43198               572.0  
43199              1298.0  

[960 rows x 5 columns]

Notice our index still shows 43199 indexed as the last value, though we only have 960 rows in the `stroke` data frame. It is good practice to reset the index after processing your data frame to avoid funky and unexpected things! We can do this with `df.reset_index()`. If we specify the argument `drop = True` the index column will be dropped, but default the original index is added as a column to the dataframe. This can be useful to keep if the index functions as a ID of sorts or knowing which row it a particular observation fell in the raw dataset is important to your analysis. 

In [44]:
stroke.reset_index() 
# notice index is now a new column so our data frame dimensions are now 960 rows x 6 columns

index FinancialYear AdmissionType       AgeGroup  \
0     7284       2013/14           All     0-44 years   
1     7285       2013/14           All    45-64 years   
2     7286       2013/14           All    65-74 years   
3     7287       2013/14           All   75plus years   
4     7288       2013/14           All  under75 years   
..     ...           ...           ...            ...   
955  43195       2022/23      Transfer    45-64 years   
956  43196       2022/23      Transfer    65-74 years   
957  43197       2022/23      Transfer   75plus years   
958  43198       2022/23      Transfer  under75 years   
959  43199       2022/23      Transfer            All   

                      Diagnosis  NumberOfDischarges  
0       Cerebrovascular Disease              1239.0  
1       Cerebrovascular Disease              6822.0  
2       Cerebrovascular Disease              6608.0  
3       Cerebrovascular Disease             14331.0  
4       Cerebrovascular Disease             14669.0  
..                          ...                 ...  
955  TIAs and related syndromes               262.0  
956  TIAs and related syndromes               300.0  
957  TIAs and related syndromes               726.0  
958  TIAs and related syndromes               572.0  
959  TIAs and related syndromes              1298.0  

[960 rows x 6 columns]

In [46]:
stroke = stroke.reset_index(drop = True)

stroke

FinancialYear AdmissionType       AgeGroup                   Diagnosis  \
0         2013/14           All     0-44 years     Cerebrovascular Disease   
1         2013/14           All    45-64 years     Cerebrovascular Disease   
2         2013/14           All    65-74 years     Cerebrovascular Disease   
3         2013/14           All   75plus years     Cerebrovascular Disease   
4         2013/14           All  under75 years     Cerebrovascular Disease   
..            ...           ...            ...                         ...   
955       2022/23      Transfer    45-64 years  TIAs and related syndromes   
956       2022/23      Transfer    65-74 years  TIAs and related syndromes   
957       2022/23      Transfer   75plus years  TIAs and related syndromes   
958       2022/23      Transfer  under75 years  TIAs and related syndromes   
959       2022/23      Transfer            All  TIAs and related syndromes   

     NumberOfDischarges  
0                1239.0  
1                6822.0  
2                6608.0  
3               14331.0  
4               14669.0  
..                  ...  
955               262.0  
956               300.0  
957               726.0  
958               572.0  
959              1298.0  

[960 rows x 5 columns]

### Task 13 

Check the data types of the remaining 5 variables and convert them to a better data type if needed. 


<details><summary style='color:lightblue'>HINT: Editing the dataframe. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

In order to actually change the data types in the original dataframe, make sure to assign it back to an object since the `astype()` functions returns a copy.

In [48]:
## your answer 
stroke['AgeGroup'].as_category()

AttributeError: 'Series' object has no attribute 'as_category'

#### Task 13 Solution 

In [47]:
stroke.dtypes

FinancialYear          object
AdmissionType          object
AgeGroup               object
Diagnosis              object
NumberOfDischarges    float64
dtype: object

In [49]:
# an elegant solution would to use our knowledge of data structures to our advantage 
## and input a dictionary!

stroke = stroke.astype({"FinancialYear": "category",
                       "AdmissionType": "category",
                       "AgeGroup": "category",
                       "Diagnosis": "category"})

stroke.dtypes

FinancialYear         category
AdmissionType         category
AgeGroup              category
Diagnosis             category
NumberOfDischarges     float64
dtype: object

### Task 14

Look at the categories within the categorical variables - is there anything unexpected?


<details><summary style='color:lightblue'>HINT: A warning about the data not to miss. CLICK HERE TO SEE THE ANSWER. BUT REALLY TRY TO DO IT YOURSELF FIRST!</summary>

It looks like some of our variables include aggregate level responses! Good thing we checked our data. Aggregate data is very common in health and social care data. It is crucial to check your data to ensure you are aware of any aggregate categories. Depending on your specific use case, you may wish to use only the aggregate levels or perhaps remove the aggregate levels and only work with the finer-grained categories.

In [ ]:
## your answer



#### Task 14 Solution 

`AdmissionType` includes 3 unique levels as well as an aggregate "All" level. 

`AgeGroup` includes 2 aggregate levels: "All" and "Under75 years" and should be ordered. 

`Diagnosis` does not have any aggregate levels.

In [50]:
# there are different syntax approach you could take here 
## both do the same things, its more a stylistic preference

stroke.AdmissionType.unique()

['All', 'Elective', 'Emergency', 'Transfer']
Categories (4, object): ['All', 'Elective', 'Emergency', 'Transfer']

In [51]:
stroke.FinancialYear.unique()

['2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23']
Categories (10, object): ['2013/14', '2014/15', '2015/16', '2016/17', ..., '2019/20', '2020/21', '2021/22', '2022/23']

In [52]:
stroke["AgeGroup"].unique()

['0-44 years', '45-64 years', '65-74 years', '75plus years', 'under75 years', 'All']
Categories (6, object): ['0-44 years', '45-64 years', '65-74 years', '75plus years', 'All', 'under75 years']

In [53]:
stroke["Diagnosis"].unique()

['Cerebrovascular Disease', 'Stroke', 'Subarachnoid Haemorrhage', 'TIAs and related syndromes']
Categories (4, object): ['Cerebrovascular Disease', 'Stroke', 'Subarachnoid Haemorrhage', 'TIAs and related syndromes']

## Task 15 

`AgeGroup` is a bit messy. It should be ordered and includes 2 aggregate categories. Remove the aggregate categories and order the remaining categories.


<details><summary style='color:lightblue'>HINT: Resulting data frame dimensions. CLICK HERE TO SEE</summary>

Your `stroke` data frame should contain 640 rows and 5 columns 


<details><summary style='color:lightblue'>HINT: A new "not" operator. CLICK HERE TO SEE</summary>

In Python is `~` is a bitwise operator for NOT

In [ ]:
## your answer 
stroke['AgeGroup'].

#### Task 15 Solution 

In [54]:
stroke = stroke.loc[~stroke['AgeGroup'].isin(["All", "under75 years"])]

# this code works with ~ and isin because isin returns a boolean Series
# so to select rows whose value is not in some value(s), we can negate the boolean Series using ~

In [55]:
# unique shows us the unique values within our column/Series 
# but oh no, did we not just remove All and under75 years? Why is it still there?
stroke.AgeGroup.unique

<bound method Series.unique of 0        0-44 years
1       45-64 years
2       65-74 years
3      75plus years
6        0-44 years
           ...     
951    75plus years
954      0-44 years
955     45-64 years
956     65-74 years
957    75plus years
Name: AgeGroup, Length: 640, dtype: category
Categories (6, object): ['0-44 years', '45-64 years', '65-74 years', '75plus years', 'All', 'under75 years']>

In [56]:
# we can see how many times the category is used in our the column/Series (i.e., how many rows)
stroke.AgeGroup.value_counts()

# While All and under75 years are still categories, they are empty, so all is well! 

AgeGroup
0-44 years       160
45-64 years      160
65-74 years      160
75plus years     160
All                0
under75 years      0
Name: count, dtype: int64

In [57]:
## now for the ordering of our category variable 
AgeGroup_order = ['0-44 years', '45-64 years', '65-74 years', '75plus years']

stroke["AgeGroup"] = stroke["AgeGroup"].cat.set_categories(AgeGroup_order, ordered = True)

stroke.AgeGroup.unique()

['0-44 years', '45-64 years', '65-74 years', '75plus years']
Categories (4, object): ['0-44 years' < '45-64 years' < '65-74 years' < '75plus years']

### Task 16 

Create a summary table with the average number of discharges with a stroke diagnosis by age group for all admissions in the financial years 2021/22 and 2022/23.

<details><summary style='color:lightblue'>HINT: Breaking down the task. CLICK HERE TO SEE</summary>

Further filtering of the data is needed for this task. THEN grouping the data in order to compute our values of interest


In [67]:
## your answer 
stroke.loc[(stroke['AdmissionType'] == 'All') &
            ((stroke['FinancialYear'] == '2021/22') | (stroke['FinancialYear'] == '2022/23')) &
            (stroke['Diagnosis'] == 'Stroke')].groupby(['FinancialYear', 'AgeGroup'])["NumberOfDischarges"].mean().dropna()

/var/folders/wd/r_s1bkq94kjf9m96tlrt6_zh0000gn/T/ipykernel_82959/1637612309.py:4: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  (stroke['Diagnosis'] == 'Stroke')].groupby(['FinancialYear', 'AgeGroup'])["NumberOfDischarges"].mean().dropna()


FinancialYear  AgeGroup    
2021/22        0-44 years        807.0
               45-64 years      5608.0
               65-74 years      6009.0
               75plus years    12973.0
2022/23        0-44 years        777.0
               45-64 years      5565.0
               65-74 years      6085.0
               75plus years    13745.0
Name: NumberOfDischarges, dtype: float64

#### Task 16 solution

In [ ]:
stroke.loc[(stroke["AdmissionType"] == "All") & 
           (stroke["Diagnosis"]== "Stroke") & 
           (stroke["FinancialYear"].isin(["2021/22", "2022/23"]))].groupby(["FinancialYear", "AgeGroup"])["NumberOfDischarges"].mean().dropna()

# change the order of the variables in the groupby list to see how the output changes 

---
## Well done! 🎉 

Well done! You have completed all of the tasks for the Python notebook for this tutorial. If you have not done so yet, now move to the R notebook.

Do not forget your 3 stars, a wish, and a step mini-diaries for this week once you have completed the tutorial notebooks and content for the week. 


---
*Dr. Brittany Blankinship (2024)*